src/lib/address_resolve/AddressResolve_DefaultImpl.cpp

```diff
@@ -0,0 +1,309 @@
+        if (ip.IsIPv6ULA())
+        {
+            return IpScore::kUniqueLocal;
+        }
+
+        if (ip.IsIPv6LinkLocal())
+        {
+            return IpScore::kLinkLocal;
+        }
+
+        return IpScore::kOtherIpv6;
+    }
+    else
+    {
+        return IpScore::kIpv4;
+    }
+}
+
+} // namespace
+
+void NodeLookupHandle::ResetForLookup(System::Clock::Timestamp now, const NodeLookupRequest & request)
+{
+    mRequestStartTime = now;
+    mRequest          = request;
+    mBestPeerAddress  = Transport::PeerAddress();
+    mBestAddressScore = ScoreValue(IpScore::kInvalid);
+}
+
+void NodeLookupHandle::LookupResult(const Transport::PeerAddress & addr)
+{
+    unsigned newScore = ScoreValue(ScoreIpAddress(addr.GetIPAddress(), addr.GetInterface()));
+    if (newScore > mBestAddressScore)
+    {
+        mBestPeerAddress  = addr;
+        mBestAddressScore = newScore;
+
+#if CHIP_PROGRESS_LOGGING
+        char addr_string[Transport::PeerAddress::kMaxToStringSize];
+        mBestPeerAddress.ToString(addr_string);
+        ChipLogProgress(Discovery, "Address %s is scored at %u", addr_string, mBestAddressScore);
+#endif
+    }
+}
+
+System::Clock::Timeout NodeLookupHandle::NextEventTimeout(System::Clock::Timestamp now)
+{
+    const System::Clock::Timestamp elapsed = now - mRequestStartTime;
+
+    if (elapsed < mRequest.GetMinLookupTime())
+    {
+        return mRequest.GetMinLookupTime() - elapsed;
+    }
+    else if (elapsed < mRequest.GetMaxLookupTime())
+    {
+        return mRequest.GetMaxLookupTime() - elapsed;
+    }
+    else
+    {
+        ChipLogError(Discovery, "Unexpected timeout: lookup should have been cleaned already.");
+        return System::Clock::Timeout::zero();
+    }
+}
+
+NodeLookupAction NodeLookupHandle::NextAction(System::Clock::Timestamp now)
+{
+    const System::Clock::Timestamp elapsed = now - mRequestStartTime;
+
+    ChipLogProgress(Discovery, "Checking node lookup status after %lu ms", static_cast<unsigned long>(elapsed.count()));
+
+    // We are still within the minimal search time. Wait for more results.
+    if (elapsed < mRequest.GetMinLookupTime())
+    {
+        ChipLogProgress(Discovery, "Keeping DNSSD lookup active");
+        return NodeLookupAction::kKeepSearching;
+    }
+
+    // Minimal time to search reached. If any Ip available, ready to return it.
+    if (mBestAddressScore > ScoreValue(IpScore::kInvalid))
+    {
+        GetListener()->OnNodeAddressResolved(GetRequest().GetPeerId(), mBestPeerAddress);
+        return NodeLookupAction::kStopSearching;
+    }
+
+    // Give up if the maximum search time has been reached
+    if (elapsed >= mRequest.GetMaxLookupTime())
+    {
+        GetListener()->OnNodeAddressResolutionFailed(GetRequest().GetPeerId(), CHIP_ERROR_TIMEOUT);
+        return NodeLookupAction::kStopSearching;
+    }
+
+    return NodeLookupAction::kKeepSearching;
+}
+
+CHIP_ERROR Resolver::LookupNode(const NodeLookupRequest & request, Impl::NodeLookupHandle & handle)
```

Comment:
Do we need another DNS-SD layer for the scoring logic? My concerns are that:
* we already have `ResolverProxy` built on top of the `Resolver` and I'm not sure how your address resolver and the `ResolverProxy` will work together.
* some platforms seem to already implement waiting for multiple results or even choosing the best address. For example, based on #15239 Darwin seems to return results to the DiscoveryImplPlatform once no more records are coming and Avahi only returns a single address (so I guess it must implement some scoring logic itself). Would be great to avoid the extra code size in such cases.

Hence I wonder if it would be possible to let the platforms choose how long they want to wait for more results and just add a new interface for address scoring strategy that the platforms could use if they support receiving multiple addresses. Or am I missing some of your requirements and it's not that simple? :)

Reply:
Long  term plan is to drop resolver proxy:
  - ResolverProxy seems to just wrap another resolver
  - ResolverProxy is circular in the PlatformDnsimplementation (found in this PR: ResolverProxy initializes PlatformDNS and PlatformDNS contains a resolver proxy).
  - I am not convinced resolver proxy handles multiple parallel lookups - it still calls into dns resolver with a single delegate, so it likely moves the delegate around at call time.

Reply:
Regarding choosing: This class can be replaced at build time using "custom" for the build implementation for extreamly granular control.

If using existing class - which includes some RAM overhead,  timeouts can be set to 0 for min wait time, in which case the first result will immediately be returned.

Reply:
For multiple parallel lookups: I believe platform dns works because it has handles, howver minmdns does not support parallel lookup with ResolverProxy. If using the updated lookup method, we both can get better results and do parallel lookups on platform and minmdns.